### PR TITLE
fixed __repr__ function of Headers and Url classes to not show auth info explicitly.

### LINKED ISSUE #220: Don't render Authorization, Proxy-Authorization headers, password in URLs
We could render them as `[secure]` when calling `__repr__` so that they're not exposed by logging services. Was wondering what others thought of this idea?

### PR DESCRIPTION
**Pull request by cansarigol**
Closes #220

### CODE REVIEW

File: httpx/models.py [hunk 1]
```diff
@@ -482,17 +486,31 @@ def __eq__(self, other: typing.Any) -> bool:
...
         class_name = self.__class__.__name__
 
         encoding_str = ""
         if self.encoding != "ascii":
             encoding_str = f", encoding={self.encoding!r}"
-
-        as_dict = dict(self.items())
-        if len(as_dict) == len(self):
+        clone = self.copy(hide_sensitive_items=True)
+        as_dict = dict(clone.items())
```
Comment (sethmlarson) [inline httpx/models.py]: What do you think about grabbing the items and filtering their value based on being included in a list:

ie:
```python
sensitive_headers = {"authorization", "proxy-authorization"}
clone = [(k, "[secure]" if k in sensitive_headers else v for k, v in self.items()]
```

Would this work the same?

File: httpx/models.py [hunk 2]
```diff
@@ -205,7 +206,10 @@ def __eq__(self, other: typing.Any) -> bool:
         return isinstance(other, (URL, str)) and str(self) == str(other)
 
     def __str__(self) -> str:
-        return self._uri_reference.unsplit()
+        url = self._uri_reference.unsplit()
+        if self._uri_reference.userinfo:
+            url = url.replace(self._uri_reference.userinfo, "[secure]")
```
Comment (sethmlarson) [inline httpx/models.py]: We might have to reimplement `rfc3986.URIReference.unsplit()` here. Also we shouldn't obscure the username field, just the password field.

**Comment by cansarigol** [comment]
Thanks for help @sethmlarson. I applied your comments

File: httpx/models.py [hunk 3]
```diff
@@ -205,6 +205,13 @@ def __eq__(self, other: typing.Any) -> bool:
...
 
     def __str__(self) -> str:
+        if self._uri_reference.userinfo:
+            userinfo = f"{self.username}:[secure]"
+            return self._uri_reference.copy_with(
+                authority=rfc3986.normalizers.normalize_authority(
+                    (userinfo, self._uri_reference.host, self._uri_reference.port)
+                )
+            ).unsplit()
         return self._uri_reference.unsplit()
```
Comment (tomchristie) [inline httpx/models.py]: The `__str__` method shouldn't obscure the URL password. We should only do that for `__repr__`.

(We still want to allow explicit `str()` casting to function normally, we just don't want credentials leaking out of variables representations in tracebacks etc.)

Comment (cansarigol) [reply httpx/models.py]: Thanks @tomchristie

File: httpx/models.py [hunk 4]
```diff
@@ -209,7 +209,19 @@ def __str__(self) -> str:
 
     def __repr__(self) -> str:
         class_name = self.__class__.__name__
-        url_str = str(self)
```
Comment (tomchristie) [inline httpx/models.py]: Let's use a proper block `if ... else ...` here. I think it's too opaque otherwise.

File: httpx/models.py [hunk 5]
```diff
@@ -489,11 +501,15 @@ def __repr__(self) -> str:
         if self.encoding != "ascii":
             encoding_str = f", encoding={self.encoding!r}"
 
-        as_dict = dict(self.items())
-        if len(as_dict) == len(self):
+        sensitive_headers = {"authorization", "proxy-authorization"}
+        clone = [
```
Comment (tomchristie) [inline httpx/models.py]: Let's call this variable `as_list` so that it mirrors the existing `as_dict`.

**Comment by tomchristie** [comment]
I'm liking this, yup! Coupla minor comments.

**Comment by cansarigol** [comment]
Thanks again @tomchristie, I applied your comments.

File: httpx/models.py [hunk 6]
```diff
@@ -209,7 +209,18 @@ def __str__(self) -> str:
 
     def __repr__(self) -> str:
         class_name = self.__class__.__name__
-        url_str = str(self)
+        if self._uri_reference.userinfo:
+            url_str = self._uri_reference.copy_with(
```
Comment (tomchristie) [inline httpx/models.py]: Can this be simplified to `self._uri_reference.copy_with(userinfo=f"{self.username}:[secure]").unsplit()`?

**Comment by tomchristie** [comment]
General thought: Do we prefer `[secure]` or `********` as the placeholder here?

We use the asterix style in Starlette, and I quite like it as being language independent, and mirroring eg. password input styling. What prior art is there in other places in Python - is there anything similar in eg. werkzeug or urllib or requests, or anything?

File: httpx/models.py (hunk 6 above)
Comment (cansarigol) [reply httpx/models.py]: as far as I understand, we just copy `scheme, authority, path, query, fragment` ([link](https://github.com/python-hyper/rfc3986/blob/2982509402519575282845adf1f39dc0195369e0/src/rfc3986/_mixin.py#L321))

**Comment by sethmlarson** [comment]
Btw I chose `[secure]` just because it's what Travis does to obscure private environment values.

File: httpx/models.py (hunk 6 above)
Comment (sethmlarson) [reply httpx/models.py]: What if `self.username` is `None`?

Comment (tomchristie) [reply httpx/models.py]: `copy_with()` supports [a `userinfo=...` argument](https://rfc3986.readthedocs.io/en/latest/user/parsing.html?highlight=userinfo#some-examples).

So, I think this section of code can be...

```python
    if self._uri_reference.userinfo:
        self._uri_reference.copy_with(userinfo=f"{self.username}:[secure]").unsplit()
```

Note that self.username cannot be `None` if some userinfo value is set.

Comment (sethmlarson) [reply httpx/models.py]: Nevermind, totally forgot that it's `self.username` not `uri_reference.username`. Carry on! :)

**Comment by StephenBrown2** [comment]
Random comment from the peanut gallery: I'm in favor of `********`, as pydantic uses asterisks for it's Secret types as well: https://pydantic-docs.helpmanual.io/#secret-types

**Comment by pquentin** [comment]
Another data point: [Sentry uses [Filtered]](https://help.sentry.io/hc/en-us/articles/115000154594-Why-am-I-seeing-Filtered-in-my-event-data-). (But `********` is fine.)

**Comment by cansarigol** [comment]
I'm using Elastic APM. it uses `********`

**Pull request merged**